### ISSUE TITLE
Build fails due to a failing test on teamcity

### ISSUE THREAD

**Issue by iqbalskd** [Starcounter/Starcounter.Nova.Cli#82]
On local it works fine but a test fails when I run it through Bifrost:

https://teamcity.starcounter.org/viewLog.html?buildId=69782&buildTypeId=Bifrost_IntNovaWinDbg&tab=buildLog

Possibly we haven't released all the depencies yet.

Bifrost Branch:

https://github.com/Starcounter/Bifrost.BuildEnvironment/tree/iqbal-issue-62

@Mackiovello

**Comment by ubbeK** [comment]
@Mackiovello Clarification.

It works when only using released nuget packages, not in the Integration build.

**Comment by Mackiovello** [comment]
> It works when only using released nuget packages, not in the Integration build.

That is surprising. I'll investigate this tomorrow. There's no change that I can remember that should break Nova.Cli.

**Comment by Mackiovello** [comment]
There seems to be some kind of regression in level0. When I run with locally built packages from level0 develop branch, two tests fail:

```
[xUnit.net 00:00:02.3386923]     Starcounter.Nova.Administration.IntegrationTests.SqlCommandTests.CanQueryDatabase [FAIL]
Failed   Starcounter.Nova.Administration.IntegrationTests.SqlCommandTests.CanQueryDatabase
Error Message:
 System.AggregateException : One or more errors occurred. (Unknown errorcode:90) (The following constructor parameters did not have matching fixture data: AppHostFixture fixture)
---- Starcounter.Nova.StarcounterException : Unknown errorcode:90
---- The following constructor parameters did not have matching fixture data: AppHostFixture fixture
Stack Trace:

----- Inner Stack Trace #1 (Starcounter.Nova.StarcounterException) -----
   at Starcounter.Nova.Hosting.Connection.Start()
   at Starcounter.Nova.Hosting.ConnectedHost.Start()
   at Starcounter.Nova.Administration.IntegrationTests.AppHostFixture..ctor() in /home/erlend/Starcounter/Starcounter.Nova.Cli/test/Starcounter.Nova.Administration.IntegrationTests/SqlCommandTests.cs:line 68
----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----
```

```

[xUnit.net 00:00:01.4435971]     Starcounter.Nova.DatabaseReader.Tests.Tests.CreatedTableModel_HasRightNumberOfColumns [FAIL]
Failed   Starcounter.Nova.DatabaseReader.Tests.Tests.CreatedTableModel_HasRightNumberOfColumns
Error Message:
 Starcounter.Nova.StarcounterException : Unknown errorcode:90
Stack Trace:
   at Starcounter.Nova.Hosting.Connection.Start()
   at Starcounter.Nova.Hosting.ConnectedHost.Start()
   at TestUtilities.TempAppHost.Start() in /home/erlend/Starcounter/Starcounter.Nova.Cli/test/TestUtilities/TempAppHost.cs:line 37
   at TestUtilities.TestAppHost..ctor() in /home/erlend/Starcounter/Starcounter.Nova.Cli/test/TestUtilities/TestAppHost.cs:line 7
   at Starcounter.Nova.DatabaseReader.Tests.DatabaseReaderTestAppHost..ctor()
```

In the directory of the test database for the `CanQueryDatabase`, there's a file "starcounter.notify" with these entries:

```
0 201 01180827T101026 
0 607 01180827T101027 
```

The Nova version doesn't matter, it's the same result no matter if you use latest develop or the NuGet packages.

I'm a little lost on what to do.  @bigwad or @linkdata do you have any idea?

On windows, one test fails ([build](https://teamcity.starcounter.org/viewLog.html?buildId=69782&buildTypeId=Bifrost_IntNovaWinDbg&tab=buildLog)):

```
System.AggregateException : One or more errors occurred. (ScErrConnConnectToDatabase (SCERR7009): Can't establish connection to specified database.) (The following constructor parameters did not have matching fixture data: AppHostFixture fixture)
---- Starcounter.Nova.StarcounterException : ScErrConnConnectToDatabase (SCERR7009): Can't establish connection to specified database.
---- The following constructor parameters did not have matching fixture data: AppHostFixture fixture

----- Inner Stack Trace #1 (Starcounter.Nova.StarcounterException) -----
   at Starcounter.Nova.Error.Check(UInt32 code)
   at Starcounter.Nova.Hosting.Connection.Start()
   at Starcounter.Nova.Hosting.ConnectedHost.Start()
   at Starcounter.Nova.Administration.IntegrationTests.AppHostFixture..ctor() in C:\tc_work\Bifrost_IntNovaWinDbg-1276\Starcounter.Nova.Cli\test\Starcounter.Nova.Administration.IntegrationTests\SqlCommandTests.cs:line 68
----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----
```

On Linux, the [integration build](https://teamcity.starcounter.org/viewLog.html?buildId=69865&buildTypeId=Bifrost_IntNovaLinDbg&tab=buildLog) fails with this:

```
[Step 1/1] /usr/share/dotnet/sdk/2.1.200/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.TargetFrameworkInference.targets(135,5): error : The current .NET SDK does not support targeting .NET Core 2.1.  Either target .NET Core 2.0 or lower, or use a version of the .NET SDK that supports .NET Core 2.1. [/home/teamcity/TeamCity/BuildAgent/work/Bifrost_IntNovaLinDbg-1247/Starcounter.Nova.Cli/src/Starcounter.Nova.Cli/Starcounter.Nova.Cli.csproj]
```

That should be relatively easy to fix by upgrading the .NET SDK on the build agents.

So to summarize:
* Succeeds when using published packages on both Linux and Windows
* Two tests fail on Linux when run locally with level0 on develop with `Unknown errorcode:90`
* Linux integration builds fail on TC because of wrong .NET SDK
* Windows integration build on TC fails on the test `CanQueryDatabase` with the exception `ScErrConnConnectToDatabase (SCERR7009)`

**Comment by Mackiovello** [comment]
I also found this in the log file:

```
01180827T103152 Error TestDatabase Starcounter 6058 ScErrUnexpErrorCreateDbMemFile (SCERR6058): An unexpected error was detected when creating a memory file for database data. OS error=17 (shm_open).
01180827T103152 Critical TestDatabase Starcounter 4186 A fatal error was detected: ScErrCantCreateDbMemoryFile (SCERR4186): The memory manager was unable to create a memory file to store database memory.
01180827T103152 Error TestDatabase Starcounter 6058 ScErrUnexpErrorCreateDbMemFile (SCERR6058): An unexpected error was detected when creating a memory file for database data. OS error=17 (shm_open).
01180827T103152 Critical TestDatabase Starcounter 4186 A fatal error was detected: ScErrCantCreateDbMemoryFile (SCERR4186): The memory manager was unable to create a memory file to store database memory.
```

**Comment by Mackiovello** [comment]
> Could be that someone ran a database under the root and now build agent account (presumably not root) can't clean up leftovers. Try to run make/linux_reset.sh under the root.

That works for me locally. Thanks for the hint! 

For the Linux builds to pass on TC, it should be enough to upgrade the .NET SDK to support .NET Core 2.1.

There's still the problem with the Windows integration build with `ScErrConnConnectToDatabase (SCERR7009)`. I'll try to replicate locally.

**Comment by bigwad** [comment]
>There's still the problem with the Windows integration build with ScErrConnConnectToDatabase (SCERR7009). 

Might be some problem with starting scdata. Do you have the log?

**Comment by Mackiovello** [comment]
> Might be some problem with starting scdata. Do you have the log?

No, unfortunately not. I'm not able to reproduce locally on Windows. 

Running a build on a different build agent to check if it's build agent related: https://teamcity.starcounter.org/viewLog.html?buildId=69896&tab=buildResultsDiv&buildTypeId=Bifrost_IntNovaWinDbg

**Comment by Mackiovello** [comment]
I still don't understand what's going on. The tests that fails `CanQueryDatabase` starts an application that creates a database, waits for the application to quit, connects to the database, and queries that database. 

If I look in the database directory of the test application, there's only a `.cfg` file in there after the test fails. I'll spend some more time to try to recreate this locally. I tried to use locally built Nova and Bluestar2 from develop, but that worked.

**Comment by Mackiovello** [comment]
Okay. I'll have to start to narrow things down. 

What we know is that it works with the Linux integration build but that it fails on the Windows integration build. It does pass on the Windows published build, so something has happened with the NuGet dependencies on Starcounter.Nova.Cli since then.

The possible dependencies are:
- Nova
- BluestarFFI
- Bluestar2
- Starcounter.QueryProcessor

There's also Starcounter.ErrorCodes, but I consider that breaking something extremely unlikely. 

The version used in the published builds are:
- Nova version 0.11.0
- BluestarFFI 4.0.0
- Bluestar2 version 2.0.0
- Starcounter.QueryProcessor 2.0.7

That leaves these commit ranges for when something could've happened:
- Nova https://github.com/Starcounter/Starcounter.Nova/commit/8f995e0f88010a1c6f49abbc0dfdc6eb5df5db5d to https://github.com/Starcounter/Starcounter.Nova/commit/4861f38ac050a3fdb046bbc32595372cd4de25b1
- BluestarFFI https://github.com/Starcounter/Starcounter.BluestarFFI/commit/df98eabcd8ac05054edda6f439e5a418150ae3a3 to https://github.com/Starcounter/Starcounter.BluestarFFI/commit/b4e88aeffabd6e9e5eaf33f3d5b6d86e65df7a3d
- Bluestar2 https://github.com/Starcounter/level0/commit/56d8259261b7f9c3e30b84f58ab0b29a8950df91 to https://github.com/Starcounter/level0/commit/ef123a57a91be8e74fdc3ed20b53a03484e1ae4a
- Starcounter.QueryProcessor https://github.com/Starcounter/level1/commit/54d6b32adb8df1a2d0bc4f652d5f98b9e5e1288c to https://github.com/Starcounter/level1/commit/ee07f74ddba864f297a8312faf3fb9255eecf097

I'll try to rule out one thing at a time. It's a little more complicated due to the breaking Bifrost change, but it's still possible.

**Comment by Mackiovello** [comment]
[Windows Nova integration build](https://teamcity.starcounter.org/viewLog.html?buildId=70108&) with level1 at https://github.com/Starcounter/level1/commit/54d6b32adb8df1a2d0bc4f652d5f98b9e5e1288c. If this build passes, Starcounter.QueryProcessor has broken something. I think that's unlikely.

[Windows Nova integration build](https://teamcity.starcounter.org/viewLog.html?buildId=70110&tab=buildResultsDiv&buildTypeId=Bifrost_IntNovaWinDbg) with BluestarFFI at https://github.com/Starcounter/Starcounter.BluestarFFI/commit/df98eabcd8ac05054edda6f439e5a418150ae3a3. If this build passes, Starcounter.BluestarFFI has broken something. Again, unlikely.

**Comment by Mackiovello** [comment]
[Windows Nova Integration build](https://teamcity.starcounter.org/viewLog.html?buildId=70111&) with Level0 at  https://github.com/Starcounter/level0/commit/56d8259261b7f9c3e30b84f58ab0b29a8950df91. If this build passes level0 has broken something. I think that's the most likely.

**Comment by Mackiovello** [comment]
[Windows Nova integration build](https://teamcity.starcounter.org/viewQueued.html?itemId=70117) with Nova at  https://github.com/Starcounter/Starcounter.Nova/commit/8f995e0f88010a1c6f49abbc0dfdc6eb5df5db5d. If this build passes, Nova has broken something.

**Comment by Mackiovello** [comment]
Here's the diff for the changes between the commits that broke this build: https://github.com/Starcounter/level0/compare/2edb433..e67ffd7

Remember that it just broke a test on Windows. The failing test is here: https://github.com/Starcounter/Starcounter.Nova.Cli/blob/develop/test/Starcounter.Nova.Administration.IntegrationTests/SqlCommandTests.cs. Do we have anyone that can look at this? The people that have contributed code during that period are @bigwad, @andan342, @henriksjostrom and Alexander.

**Comment by henriksjostrom** [comment]
@Mackiovello Is it possible for you to bisect it to further reduce the potential sources of the issue?

**Comment by Mackiovello** [comment]
> @Mackiovello Is it possible for you to bisect it to further reduce the potential sources of the issue?

Sure, I can do that.

**Comment by bigwad** [comment]
@Mackiovello, I see this issue is not 100% reproducible. Right?

**Comment by Mackiovello** [comment]
> @Mackiovello, I see this issue is not 100% reproducible. Right?

Yes, it's always reproducible. What makes you think that it's not?

**Comment by bigwad** [comment]
>What makes you think that it's not?

build history. most of the builds succeed.

**Comment by Mackiovello** [comment]
> build history. most of the builds succeed.

Yes, the builds with the level0 branch set to a branch where the mentioned commits aren't included succeed. Look at the level0 branch parameter of the different builds. They should use a variant of the branch issue/cli/52-[n].

**Comment by bigwad** [comment]
Here is what going on - SQLTest.core and Starcounter.Nova.Cli use different versions of bluestar2, which turned out to be runtime incompatible if run side-by-side. I don't think, it's critical, but will fix it.
What's more intriguing - is why they use different versions in integration build, which is supposed to use only locally built bluestar.

**Comment by Mackiovello** [comment]
Thanks for the investigation @bigwad!

> What's more intriguing - is why they use different versions in integration build, which is supposed to use only locally built bluestar.

Yes, this is a problem. It's tracked here: https://github.com/Starcounter/AdminTrack/issues/564

> Also, it's not gonna work in release due to this line:

I'll fix that. Thanks for catching it. 

> Putting aside the method used to create a database - there is a severe problem with ignoring process error codes here:

Can you suggest how it should be done?

**Comment by Mackiovello** [comment]
Thanks to [the fix](https://github.com/Starcounter/level0/pull/1151) by @bigwad, the windows build now passes: https://teamcity.starcounter.org/viewLog.html?buildId=70787&buildTypeId=Bifrost_IntNovaWinDbg&tab=buildResultsDiv

Linux integration build still fails.: https://teamcity.starcounter.org/viewType.html?buildTypeId=Bifrost_IntNovaLinDbg. We need to upgrade the .NET Core SDK on the build agents to support .NET Core 2.1 for this to work @iqbalskd.

With that, we no longer need to update Bluestar2 version in SqlTest.core: https://github.com/Starcounter/level1/pull/4785. Although, that will still be required once https://github.com/Starcounter/AdminTrack/issues/564 in place. So I'll go ahead with it anyway.

Two other takeaways:
- Make sure integration tests output process output properly: https://github.com/Starcounter/Starcounter.Nova.Cli/issues/92
- Make tests work with release configuration as well: https://github.com/Starcounter/Starcounter.Nova.Cli/issues/90

**Comment by ubbeK** [comment]
ScBuilLin01 has now latest .net core, testing build: https://teamcity.starcounter.org/viewLog.html?buildId=70922&buildTypeId=Bifrost_IntNovaLinRel

**Comment by ubbeK** [comment]
And for ScBuildLin02

* https://teamcity.starcounter.org/viewLog.html?buildId=70942&buildTypeId=Bifrost_IntNovaLinDbg
* https://teamcity.starcounter.org/viewQueued.html?itemId=70946:0:27
* https://teamcity.starcounter.org/viewQueued.html?itemId=70945:0:27
* https://teamcity.starcounter.org/viewQueued.html?itemId=70944:0:27

**Comment by ubbeK** [comment]
Both agents have been upgraded and tested. @iqbalskd you may now continue.

**Comment by iqbalskd** [comment]
Upgraded .net version on `ScBuildLin01` and `ScBuildLin02`

Closing this as done.

I will add `Starcounter.Nova.Cli` in Nova Build using this issue:
https://github.com/Starcounter/Starcounter.Nova.Cli/issues/62
